8Knot/queries/user_groups_query.py

```diff
@@ -0,0 +1,146 @@
+    # returns True if connection succeeds.
+    users_cache.ping()
+
+    # check if user is in sessions
+    if not users_cache.exists(user_id):
+        raise Exception("Expected user data under user_id not in cache.")
+    else:
+        user = json.loads(users_cache.get(user_id))
+
+    # query groups and options from Augur
+    users_groups, users_options = get_user_groups(user["username"], user["access_token"])
+
+    # stores groups and options in cache
+    groups_set = users_cache.set(name=f"{user_id}_groups", value=json.dumps(users_groups))
+    options_set = users_cache.set(name=f"{user_id}_group_options", value=json.dumps(users_options))
+
+    # returns success of operation
+    return bool(groups_set and options_set)
+
+
+def get_user_groups(username, bearer_token):
+    """Requests all user-level groups from augur frontend.
+
+    Args:
+        username (str): client's username
+        bearer_token (str): client's bearer token
+
+    Returns:
+        dict{group_name: [repo_ids]}: dict of users groups
+        list[{group_name, group_label}]: list of dicts to translate group labels to their values.
+    """
+
+    # request to get user's groups
+    augur_users_groups = augur.make_user_request(access_token=bearer_token)
+
+    # structure of the incoming data
+    # [{group_name: {favorited: False, repos: [{repo_git: asd;lfkj, repo_id=46555}, ...]}, ...]
+    # creates the group_name->repo_list mapping and the searchbar options for augur user groups
+    users_groups = {}
+    users_group_options = []
+    g = augur_users_groups.get("data")
+
+    # each of the augur user groups
+    for entry in g:
+        # group has one key- the name.
+        group_name: str = list(entry.keys())[0]
+
+        # only one value per entry- {favorited: ..., repos: ...},
+        # get the value component
+        repo_list = list(entry.values())[0]["repos"]
+
+        ids = parse_repolist(repo_list)
+
+        # don't accept empty groups
+        if len(ids) == 0:
+            continue
+
+        # using lower_name for convenience later- no .lower() calls
+        lower_name = group_name.lower()
+
+        # group_name->repo_list mapping
+        users_groups[lower_name] = ids
+
+        # searchbar options
+        # user's groups are prefixed w/ username to guarantee uniqueness in searchbar
+        users_group_options.append({"value": lower_name, "label": f"{username}: {group_name}"})
+
+    return users_groups, users_group_options
+
+
+def parse_repolist(repo_list, prepend_to_url=""):
+    """Converts repo_git URLs to
+    indexed repo_ids from startup for consumption
+    by group_name->list_of_repo_ids.
+
+    Args:
+        repo_list ([{repo_info}]): list of repo metadata from augur frontend.
+        prepend_to_url (str): string to prepend, e.g. "https://" if known not available
+
+    Returns:
+        [int]: list of repo ids identified.
+    """
+    # structure of the incoming data
```

Comment:
Why is this in comments and not doc string above?

Reply:
Because it's an implementation detail not an API detail.